desc/objectives/_coils.py

```diff
@@ -0,0 +1,504 @@
+        be flattened according to the number of inputs.
+    bounds : tuple of {float, ndarray}, optional
+        Lower and upper bounds on the objective. Overrides target.
+        Both bounds must be broadcastable to to Objective.dim_f
+    weight : {float, ndarray}, optional
+        Weighting to apply to the Objective, relative to other Objectives.
+        Must be broadcastable to to Objective.dim_f
+    normalize : bool, optional
+        Whether to compute the error in physical units or non-dimensionalize.
+    normalize_target : bool, optional
+        Whether target and bounds should be normalized before comparing to computed
+        values. If `normalize` is `True` and the target is in physical units,
+        this should also be set to True.
+        be set to True.
+    loss_function : {None, 'mean', 'min', 'max'}, optional
+        Loss function to apply to the objective values once computed. This loss function
+        is called on the raw compute value, before any shifting, scaling, or
+        normalization. Note: Has no effect for this objective.
+    deriv_mode : {"auto", "fwd", "rev"}
+        Specify how to compute jacobian matrix, either forward mode or reverse mode AD.
+        "auto" selects forward or reverse mode based on the size of the input and output
+        of the objective. Has no effect on self.grad or self.hess which always use
+        reverse mode and forward over reverse mode respectively.
+    grid : Grid, optional
+        Collocation grid containing the nodes to evaluate at.
+    name : str, optional
+        Name of the objective function.
+    """
+
+    def __init__(
+        self,
+        coils,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        loss_function=None,
+        deriv_mode="auto",
+        grid=None,
+        name=None,
+    ):
+
+        self._coils = coils
+        if target is None and bounds is None:
+            target = 2 * np.pi
+
+        super().__init__(
+            coils,
+            ["length"],
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            grid=grid,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        from desc.coils import CoilSet
+
+        super().build(use_jit=use_jit, verbose=verbose)
+
+        if self._normalize:
+            self._normalization = self._scales["a"]
+
+        self._dim_f = len(self._coils.coils) if isinstance(self._coils, CoilSet) else 1
```

Comment:
if `self._coils` is a `MixedCoilSet` made up of other `CoilSet`s then `len(self._coils.coils)` won't give you the number of actual coils.

It might be better to either use the tree flatten thing again here, or just record the length of the flattened coils in the superclass build and reference that here

Reply:
That's true. I think using `self._dim_f = len(self._grid)` would accomplish this too, right? 

But also, I'm confused because this passes the nested coils test, unless `nested_coils = MixedCoilSet.from_symmetry(mixed_coils, NFP=4)` doesn't actually create nested coils?

Reply:
yeah `from_symmetry` i think flattens everything down to a single list: https://github.com/PlasmaControl/DESC/blob/acba92f229c4a28134792707092b588fe56df71b/desc/coils.py#L1052-L1074 To properly test it you might need to do `MixedCoilSet(coilset, coilset)`. It's fine if they're the same coilset without any shift etc, since all we're testing is length/cuvature/torsion which are local to each coil.

Reply:
Okay you're right. So now, with `MixedCoilSet(coils, coils)`, I run into a problem in `compute()`.  Since the default `self._grid` is created based on the flattened `coils`, the grid won't match up with what's required for `self.things`.  e.g. if you have a nested coil set that's like `[CoilSet(n=2), CoilSet(n=2)]` then I think the only grid that works is a list of 2 grids or a single grid. A list of 4 grids doesn't seem to work but maybe I'm doing something wrong

Reply:
I think for the case where the default grid is used you can use tree map over self.coils instead of over the flattened list and that should make sure that it has the right structure

For the case of a user supplied list of grid I think there's another tree util to convert a list to a given tree structure

Reply:
By using tree map over self.coils, do you mean when getting the transforms? Or within `if self._grid is None`?

Reply:
Both

Reply:
```python
self._grid = tree_map(
                lambda x: LinearGrid(
                    N=2 * x.N + 5, NFP=getattr(x, "NFP", 1), endpoint=False
                ),
                self.things[0],
                is_leaf=lambda x: isinstance(x, _Coil) and not isinstance(x, CoilSet),
            )
print(self._grid)```

For whatever reason, this returns `self.things[0]` and not a list of `LinearGrid`. Is this some weird jax thing that I'm not understanding?

Reply:
Ahhh I think I see the issue. It's actually returning a `MixedCoilSet` where instead of coils it's made up of grids, because the output of `tree_map` has the same structure as the input. Let me think about this.

Reply:
ohhh yeah you're right. lol
<img width="713" alt="image" src="https://github.com/PlasmaControl/DESC/assets/60586957/f09d09c9-430c-46eb-bb01-b4fa90927b99">

Reply:
ok I think this works:
```python

def get_grid(coilset):
    if isinstance(coilset, desc.coils.MixedCoilSet):
        return [get_grid(x) for x in coilset]
    elif isinstance(coilset, desc.coils.CoilSet): # must be regular coilset
        return get_grid(coilset.coils[0]) # use same grid for all members
    else: # must be base coil
        return LinearGrid(
                    N=2 * coilset.N + 5, NFP=getattr(coilset, "NFP", 1), endpoint=False
                )

self._grid = get_grid(self.things[0])

```

the same basic approach should work for transforms

Reply:
wow that is so smart. I'll try it

Reply:
Okay it's a mess but it's working